storey/flow.py

```diff
@@ -187,16 +204,28 @@ def _get_recovery_step(self, exception):
         else:
             return self._recovery_step
 
-    def run(self):
+    def run(self, visited=None):
         if not self._legal_first_step and not self._runnable:
             raise ValueError("Flow must start with a source")
+
+        # Initialize visited set once at the top (only for the root call)
+        if visited is None:
+            visited = set()
+
+        # Detect cycles: if we've already visited this step, don't run it again
+        if self in visited:
+            return None
```

Comment:
Should be consistent with the rest of the function:
```suggestion
            return []
```